aten/src/ATen/native/cuda/TensorFactories.cu

```diff
@@ -26,4 +35,37 @@ Tensor& eye_out_cuda(Tensor& result, int64_t n, int64_t m) {
   return result;
 }
 
+Tensor& randperm_out_cuda(Tensor& result, int64_t n, Generator* generator) {
+  if (n < 0) {
+    std::ostringstream oss;
+    oss << "n must be non-negative, got " << n;
+    throw std::runtime_error(oss.str());
+  }
+
+  result.resize_({n});
+
+  if (n < 30000) {  // For small inputs, we offload it to CPU instead.
+    auto result_cpu = result.type().toBackend(kCPU).tensor({n});
+    randperm_out(result_cpu, n, generator);
+    result = result.type().copy(result_cpu);
```

Comment:
We should probably also check that the scalar type isn't `half` in here

Reply:
Should we throw an error when `result.type()` is `half`? I saw the following when testing:
```
>>> import torch
>>> cuda = torch.device('cuda:0')
>>> res = torch.cuda.HalfTensor()
>>> torch.randperm(100, out=res, device=cuda)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: randperm_out is not implemented for type torch.HalfTensor

>>> torch.randperm(1000000, out=res, device=cuda)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: expected scalar type Long but found Half

>>> res2 = torch.HalfTensor()
>>> torch.randperm(100, out=res2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: randperm_out is not implemented for type torch.HalfTensor
```

Reply:
@apaszke Wondering what's the right change to make here?

Reply:
We surely want to support randperm for half on CUDA, but we don't want to do that for CPU, so we might need to generate it in float on CPU, and then do a copy to a half CUDA tensor.